# Read data files - `pandas`

This notebook demonstrates the `read_csv`, `read_excel`, and `read_json` functions from pandas.

## Contents
1. Setup
1. `read_csv` & fixing errors
1. `read_excel`
1. `read_json`

## 1. Setup

Load the `pandas` Python library.

In [6]:
%python
import pandas as pd

In [7]:
%sh ls -hot /dbfs/mnt/datalab-datasets/538_data/mad-men

total 0
-rw-r--r-- 1 root 8.0K Sep 18 11:46 performer-scores.csv
-rw-r--r-- 1 root 1.9K Sep 18 11:46 README.md
-rw-r--r-- 1 root 35K Sep 18 11:46 show-data.csv

In [8]:
%sh ls -hot /dbfs/mnt/datalab-datasets/538_data/mad-men/*

-rw-r--r-- 1 root 8.0K Sep 18 11:46 /dbfs/mnt/datalab-datasets/538_data/mad-men/performer-scores.csv
-rw-r--r-- 1 root 1.9K Sep 18 11:46 /dbfs/mnt/datalab-datasets/538_data/mad-men/README.md
-rw-r--r-- 1 root 35K Sep 18 11:46 /dbfs/mnt/datalab-datasets/538_data/mad-men/show-data.csv

Notice that the output above changes if the asterisk is removed/included in the filepath of the `ls` command. 

The asterisk is required in order to produce a vector of full filepaths.

## 2. `read_csv` & fixing errors

Attempt to read the `show-data.csv` file with the pandas `read_csv` function. 

It produces an error, which we will fix below.

In [12]:
%python # error encountered
mad_men_show_data_pdf = pd.read_csv('/dbfs/mnt/datalab-datasets/538_data/mad-men/show-data.csv')
mad_men_show_data_pdf

--------------------------------------------------------------------------- 
 UnicodeDecodeError Traceback (most recent call last)
 <command-569009> in <module> () 
 1 # error encountered 
 ----> 2 mad_men_show_data_pdf = pd . read_csv ( '/dbfs/mnt/datalab-datasets/538_data/mad-men/show-data.csv' ) 
 3 mad_men_show_data_pdf 

 /databricks/python/lib/python3.5/site-packages/pandas/io/parsers.py in parser_f (filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, escapechar, comment, encoding, dialect, tupleize_cols, error_bad_lines, warn_bad_lines, skipfooter, skip_footer, doublequote, delim_whitespace, as_recarray, compact_ints, use_unsigned, low_memory, buffer_lines, memory_map, float_precision) 
 644 skip_blank_lines=skip_blank_lines)
 645 
 --> 646 return _read ( filepath_or_buffer , kwds ) 
 647 
 648 parser_f . __name__ = name 

 /databricks/python/lib/python3.5/site-packages/pandas/io/parsers.py in _read (filepath_or_buffer, kwds) 
 399 return parser 
 400 
 --> 401 data = parser . read ( ) 
 402 parser . close ( ) 
 403 return data 

 /databricks/python/lib/python3.5/site-packages/pandas/io/parsers.py in read (self, nrows) 
 937 raise ValueError ( 'skipfooter not supported for iteration' ) 
 938 
 --> 939 ret = self . _engine . read ( nrows ) 
 940 
 941 if self . options . get ( 'as_recarray' ) : 

 /databricks/python/lib/python3.5/site-packages/pandas/io/parsers.py in read (self, nrows) 
 1506 def read ( self , nrows = None ) : 
 1507 try : 
 -> 1508 data = self . _reader . read ( nrows ) 
 1509 except StopIteration : 
 1510 if self . _first_chunk : 

 /databricks/python/lib/python3.5/site-packages/pandas/parser.cpython-35m-x86_64-linux-gnu.so in pandas.parser.TextReader.read (pandas/parser.c:9977) () 

 /databricks/python/lib/python3.5/site-packages/pandas/parser.cpython-35m-x86_64-linux-gnu.so in pandas.parser.TextReader._read_low_memory (pandas/parser.c:10235) () 

 /databricks/python/lib/python3.5/site-packages/pandas/parser.cpython-35m-x86_64-linux-gnu.so in pandas.parser.TextReader._read_rows (pandas/parser.c:11254) () 

 /databricks/python/lib/python3.5/site-packages/pandas/parser.cpython-35m-x86_64-linux-gnu.so in pandas.parser.TextReader._convert_column_data (pandas/parser.c:12629) () 

 /databricks/python/lib/python3.5/site-packages/pandas/parser.cpython-35m-x86_64-linux-gnu.so in pandas.parser.TextReader._convert_tokens (pandas/parser.c:13505) () 

 /databricks/python/lib/python3.5/site-packages/pandas/parser.cpython-35m-x86_64-linux-gnu.so in pandas.parser.TextReader._convert_with_dtype (pandas/parser.c:15381) () 

 /databricks/python/lib/python3.5/site-packages/pandas/parser.cpython-35m-x86_64-linux-gnu.so in pandas.parser.TextReader._string_convert (pandas/parser.c:15591) () 

 /databricks/python/lib/python3.5/site-packages/pandas/parser.cpython-35m-x86_64-linux-gnu.so in pandas.parser._string_box_utf8 (pandas/parser.c:21111) () 

 UnicodeDecodeError : 'utf-8' codec can't decode byte 0xd5 in position 11: invalid continuation byte

Notice the error above is also repeat below when simply displaying the file with the `cat` command.

In [14]:
%sh # error encountered
cat /dbfs/mnt/datalab-datasets/538_data/mad-men/show-data.csv

--------------------------------------------------------------------------- 
 UnicodeDecodeError Traceback (most recent call last)
 <command-569018> in <module> () 
 17 if returncode != 0 and False : 
 18 raise subprocess . CalledProcessError ( returncode , cmd ) 
 ---> 19 ____databricks_percent_sh ( ) 
 20 finally : 
 21 del ____databricks_percent_sh 

 <command-569018> in ____databricks_percent_sh () 
 11 stdout = subprocess . PIPE , 
 12 universal_newlines = True)
 ---> 13 for line in iter ( p . stdout . readline , '' ) : 
 14 sys . stdout . write ( line ) 
 15 sys . stdout . flush ( ) 

 /databricks/python/lib/python3.5/codecs.py in decode (self, input, final) 
 319 # decode input (taking the buffer into account) 
 320 data = self . buffer + input 
 --> 321 ( result , consumed ) = self . _buffer_decode ( data , self . errors , final ) 
 322 # keep undecoded input until the next call 
 323 self . buffer = data [ consumed : ] 

 UnicodeDecodeError : 'utf-8' codec can't decode byte 0xd5 in position 3032: invalid continuation byte

In this case, when encountering an unknown error, I search for posts about the error on Google. 

Search for: "read_csv pandas UnicodeDecodeError utf-8 codec can't decode byte  in position invalid continuation byte". 

Notice I added "read_csv" and "pandas" to the search string. 
I removed `0xd5` and `3032` as these are specific to our file and won't help the search find other posts that solve the same problem, but with a different file. 

The search found: https://stackoverflow.com/questions/18171739/unicodedecodeerror-when-reading-csv-file-in-pandas-with-python

This links suggests trying `encoding = "ISO-8859-1"` as a parameter to `read_csv` (it works, see below)

In [16]:
%python
mad_men_show_data_pdf = pd.read_csv('/dbfs/mnt/datalab-datasets/538_data/mad-men/show-data.csv', 
                                    encoding = 'ISO-8859-1')
mad_men_show_data_pdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 248 entries, 0 to 247
Data columns (total 15 columns):
Performer 248 non-null object
Show 248 non-null object
Show Start 248 non-null int64
Show End 248 non-null object
Status? 248 non-null object
CharEnd 248 non-null int64
Years Since 248 non-null int64
#LEAD 248 non-null int64
#SUPPORT 248 non-null int64
#Shows 248 non-null int64
Score 248 non-null float64
Score/Y 248 non-null object
lead_notes 89 non-null object
support_notes 135 non-null object
show_notes 138 non-null object
dtypes: float64(1), int64(6), object(8)
memory usage: 29.1+ KB

The command did work. 

Notice the column names have non-alphabetic characters (including spaces).
The next step is to read in the columns with better names.

The names parameter specifies the column names to use (in the dataframe returned by `read_csv`).

In [19]:
%python
mad_men_show_data_pdf = pd.read_csv('/dbfs/mnt/datalab-datasets/538_data/mad-men/show-data.csv', 
                                    names=['performer', 'show', 'show_start', 'show_end',
                                           'status', 'char_end', 'years_since', 'lead', 
                                           'support', 'shows', 'score', 'score_y',
                                           'lead_notes', 'support_notes', 'show_notes'],
                                    header=0, 
                                    encoding = 'ISO-8859-1')
mad_men_show_data_pdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 248 entries, 0 to 247
Data columns (total 15 columns):
performer 248 non-null object
show 248 non-null object
show_start 248 non-null int64
show_end 248 non-null object
status 248 non-null object
char_end 248 non-null int64
years_since 248 non-null int64
lead 248 non-null int64
support 248 non-null int64
shows 248 non-null int64
score 248 non-null float64
score_y 248 non-null object
lead_notes 89 non-null object
support_notes 135 non-null object
show_notes 138 non-null object
dtypes: float64(1), int64(6), object(8)
memory usage: 29.1+ KB

Notice the column names are better, but that we needed to specify, with the `header` parameter, that row `0` containes the column names.

The column types are in the right hand column of the output from the `info` method.
If a column is read in as a `float64` or `int64` then we can assume it was read correctly. 

To check for columns that were not read correctly we only need to select columns with `object` type. 
The next command does this.

In [22]:
mad_men_show_data_pdf \
  .select_dtypes(include=['object']) \
  .head()

Out[ 32 ]: 
 performer show show_end status score_y \
0 Steven Hill Law & Order 2010 END 0 
1 Kelli Williams The Practice 2014 END 0.520833333 
2 LisaGay Hamilton The Practice 2014 END 0.333333333 
3 Lara Flynn Boyle The Practice 2014 END 0 
4 Dylan McDermott The Practice 2014 END 0.886363636 

 lead_notes \
0 NaN 
1 NaN 
2 Life of a King, 2014; Go For Sisters, 2013 
3 NaN 
4 Olympus Has Fallen, 2013; Freezer, 2014 

 support_notes \
0 NaN 
1 Any Day Now (2012) 
2 NaN 
3 NaN 
4 The Messengers, 2007; Unbeatable Harold, 2009;... 

 show_notes 
0 NaN 
1 Medical Investigation, Season 1; Lie To Me, Se... 
2 Men of a Certain Age, Season 1-2 
3 NaN 
4 Big Shots, Season 1; Dark Blue, Season 1-2; Am...

Notice that `show_end` and `score_y` should be numbers.

The following command specifies both `show_start` and `show_end` as integers. We leave `score_y` until later.

In [24]:
%python # error encountered
import numpy as np
mad_men_show_data_pdf = pd.read_csv('/dbfs/mnt/datalab-datasets/538_data/mad-men/show-data.csv', 
                                    names=['performer', 'show', 'show_start', 'show_end',
                                             'status', 'char_end', 'years_since', 'lead', 
                                             'support', 'shows', 'score', 'score_y',
                                             'lead_notes', 'support_notes', 'show_notes'],
                                    dtype={'show_start': np.int64, 'show_end': np.int64},
                                    header=0, 
                                    encoding = 'ISO-8859-1')
mad_men_show_data_pdf.info()

--------------------------------------------------------------------------- 
 ValueError Traceback (most recent call last)
 <command-575162> in <module> () 
 8 dtype = { 'show_start' : np . int64 , 'show_end' : np . int64 } , 
 9 header = 0 , 
 ---> 10 encoding = 'ISO-8859-1')
 11 mad_men_show_data_pdf . info ( ) 

 /databricks/python/lib/python3.5/site-packages/pandas/io/parsers.py in parser_f (filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, escapechar, comment, encoding, dialect, tupleize_cols, error_bad_lines, warn_bad_lines, skipfooter, skip_footer, doublequote, delim_whitespace, as_recarray, compact_ints, use_unsigned, low_memory, buffer_lines, memory_map, float_precision) 
 644 skip_blank_lines=skip_blank_lines)
 645 
 --> 646 return _read ( filepath_or_buffer , kwds ) 
 647 
 648 parser_f . __name__ = name 

 /databricks/python/lib/python3.5/site-packages/pandas/io/parsers.py in _read (filepath_or_buffer, kwds) 
 399 return parser 
 400 
 --> 401 data = parser . read ( ) 
 402 parser . close ( ) 
 403 return data 

 /databricks/python/lib/python3.5/site-packages/pandas/io/parsers.py in read (self, nrows) 
 937 raise ValueError ( 'skipfooter not supported for iteration' ) 
 938 
 --> 939 ret = self . _engine . read ( nrows ) 
 940 
 941 if self . options . get ( 'as_recarray' ) : 

 /databricks/python/lib/python3.5/site-packages/pandas/io/parsers.py in read (self, nrows) 
 1506 def read ( self , nrows = None ) : 
 1507 try : 
 -> 1508 data = self . _reader . read ( nrows ) 
 1509 except StopIteration : 
 1510 if self . _first_chunk : 

 /databricks/python/lib/python3.5/site-packages/pandas/parser.cpython-35m-x86_64-linux-gnu.so in pandas.parser.TextReader.read (pandas/parser.c:9977) () 

 /databricks/python/lib/python3.5/site-packages/pandas/parser.cpython-35m-x86_64-linux-gnu.so in pandas.parser.TextReader._read_low_memory (pandas/parser.c:10235) () 

 /databricks/python/lib/python3.5/site-packages/pandas/parser.cpython-35m-x86_64-linux-gnu.so in pandas.parser.TextReader._read_rows (pandas/parser.c:11254) () 

 /databricks/python/lib/python3.5/site-packages/pandas/parser.cpython-35m-x86_64-linux-gnu.so in pandas.parser.TextReader._convert_column_data (pandas/parser.c:12629) () 

 /databricks/python/lib/python3.5/site-packages/pandas/parser.cpython-35m-x86_64-linux-gnu.so in pandas.parser.TextReader._convert_tokens (pandas/parser.c:13876) () 

 ValueError : invalid literal for int() with base 10: 'PRESENT'

An error is encountered, indicating that `PRESENT` is an invalid value for an integer. 

So we specify `PRESENT` to indicate a missing value. This should work.

In [26]:
%python
import numpy as np
mad_men_show_data_pdf = pd.read_csv('/dbfs/mnt/datalab-datasets/538_data/mad-men/show-data.csv', 
                                    names=['performer', 'show', 'show_start', 'show_end',
                                             'status', 'char_end', 'years_since', 'lead', 
                                             'support', 'shows', 'score', 'score_y',
                                             'lead_notes', 'support_notes', 'show_notes'],
                                    dtype={'show_start': np.int64, 'show_end': np.int64, 'score_y': np.float64},
                                    header=0, 
                                    na_values=['PRESENT'],
                                    encoding = 'ISO-8859-1')
mad_men_show_data_pdf.info()

--------------------------------------------------------------------------- 
 ValueError Traceback (most recent call last)
 <command-575165> in <module> () 
 8 header = 0 , 
 9 na_values = [ 'PRESENT' ] , 
 ---> 10 encoding = 'ISO-8859-1')
 11 mad_men_show_data_pdf . info ( ) 

 /databricks/python/lib/python3.5/site-packages/pandas/io/parsers.py in parser_f (filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, escapechar, comment, encoding, dialect, tupleize_cols, error_bad_lines, warn_bad_lines, skipfooter, skip_footer, doublequote, delim_whitespace, as_recarray, compact_ints, use_unsigned, low_memory, buffer_lines, memory_map, float_precision) 
 644 skip_blank_lines=skip_blank_lines)
 645 
 --> 646 return _read ( filepath_or_buffer , kwds ) 
 647 
 648 parser_f . __name__ = name 

 /databricks/python/lib/python3.5/site-packages/pandas/io/parsers.py in _read (filepath_or_buffer, kwds) 
 399 return parser 
 400 
 --> 401 data = parser . read ( ) 
 402 parser . close ( ) 
 403 return data 

 /databricks/python/lib/python3.5/site-packages/pandas/io/parsers.py in read (self, nrows) 
 937 raise ValueError ( 'skipfooter not supported for iteration' ) 
 938 
 --> 939 ret = self . _engine . read ( nrows ) 
 940 
 941 if self . options . get ( 'as_recarray' ) : 

 /databricks/python/lib/python3.5/site-packages/pandas/io/parsers.py in read (self, nrows) 
 1506 def read ( self , nrows = None ) : 
 1507 try : 
 -> 1508 data = self . _reader . read ( nrows ) 
 1509 except StopIteration : 
 1510 if self . _first_chunk : 

 /databricks/python/lib/python3.5/site-packages/pandas/parser.cpython-35m-x86_64-linux-gnu.so in pandas.parser.TextReader.read (pandas/parser.c:9977) () 

 /databricks/python/lib/python3.5/site-packages/pandas/parser.cpython-35m-x86_64-linux-gnu.so in pandas.parser.TextReader._read_low_memory (pandas/parser.c:10235) () 

 /databricks/python/lib/python3.5/site-packages/pandas/parser.cpython-35m-x86_64-linux-gnu.so in pandas.parser.TextReader._read_rows (pandas/parser.c:11254) () 

 /databricks/python/lib/python3.5/site-packages/pandas/parser.cpython-35m-x86_64-linux-gnu.so in pandas.parser.TextReader._convert_column_data (pandas/parser.c:12629) () 

 /databricks/python/lib/python3.5/site-packages/pandas/parser.cpython-35m-x86_64-linux-gnu.so in pandas.parser.TextReader._convert_tokens (pandas/parser.c:13296) () 

 /databricks/python/lib/python3.5/site-packages/pandas/parser.cpython-35m-x86_64-linux-gnu.so in pandas.parser.TextReader._convert_with_dtype (pandas/parser.c:14304) () 

 ValueError : Integer column has NA values in column 3

But this does not work. So we check Google again for the error.

Searching for "ValueError: Integer column has NA values in column 3" yields this link:
- https://stackoverflow.com/questions/21287624/convert-pandas-column-containing-nans-to-dtype-int

The post indicates that integers cannot have missing values in pandas and that they should be retyped to `np.float32`. 
Both columns (`show_start` and `show_end`) where coding this way for consistency. 
It's awkward, but it works.

In [29]:
%python
import numpy as np
mad_men_show_data_pdf = pd.read_csv('/dbfs/mnt/datalab-datasets/538_data/mad-men/show-data.csv', 
                                    names=['performer', 'show', 'show_start', 'show_end',
                                           'status', 'char_end', 'years_since', 'lead', 
                                           'support', 'shows', 'score', 'score_y',
                                           'lead_notes', 'support_notes', 'show_notes'],
                                    dtype={'show_start': np.float32, 'show_end': np.float32},
                                    header=0, 
                                    na_values=['PRESENT'],
                                    encoding = 'ISO-8859-1')
mad_men_show_data_pdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 248 entries, 0 to 247
Data columns (total 15 columns):
performer 248 non-null object
show 248 non-null object
show_start 248 non-null float32
show_end 219 non-null float32
status 248 non-null object
char_end 248 non-null int64
years_since 248 non-null int64
lead 248 non-null int64
support 248 non-null int64
shows 248 non-null int64
score 248 non-null float64
score_y 248 non-null object
lead_notes 89 non-null object
support_notes 135 non-null object
show_notes 138 non-null object
dtypes: float32(2), float64(1), int64(5), object(7)
memory usage: 27.2+ KB

Recall that`score_y` should be a `float`. The command below specifies this in the `dtype` parameter.

In [31]:
%python
import numpy as np
mad_men_show_data_pdf = pd.read_csv('/dbfs/mnt/datalab-datasets/538_data/mad-men/show-data.csv', 
                                    names=['performer', 'show', 'show_start', 'show_end',
                                           'status', 'char_end', 'years_since', 'lead', 
                                           'support', 'shows', 'score', 'score_y',
                                           'lead_notes', 'support_notes', 'show_notes'],
                                    dtype={'show_start': np.float32, 
                                           'show_end': np.float32,
                                           'score_y': np.float32},
                                    header=0, 
                                    na_values=['PRESENT'],
                                    encoding = 'ISO-8859-1')
mad_men_show_data_pdf.info()

--------------------------------------------------------------------------- 
 ValueError Traceback (most recent call last)
 <command-858154> in <module> () 
 10 header = 0 , 
 11 na_values = [ 'PRESENT' ] , 
 ---> 12 encoding = 'ISO-8859-1')
 13 mad_men_show_data_pdf . info ( ) 

 /databricks/python/lib/python3.5/site-packages/pandas/io/parsers.py in parser_f (filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, escapechar, comment, encoding, dialect, tupleize_cols, error_bad_lines, warn_bad_lines, skipfooter, skip_footer, doublequote, delim_whitespace, as_recarray, compact_ints, use_unsigned, low_memory, buffer_lines, memory_map, float_precision) 
 644 skip_blank_lines=skip_blank_lines)
 645 
 --> 646 return _read ( filepath_or_buffer , kwds ) 
 647 
 648 parser_f . __name__ = name 

 /databricks/python/lib/python3.5/site-packages/pandas/io/parsers.py in _read (filepath_or_buffer, kwds) 
 399 return parser 
 400 
 --> 401 data = parser . read ( ) 
 402 parser . close ( ) 
 403 return data 

 /databricks/python/lib/python3.5/site-packages/pandas/io/parsers.py in read (self, nrows) 
 937 raise ValueError ( 'skipfooter not supported for iteration' ) 
 938 
 --> 939 ret = self . _engine . read ( nrows ) 
 940 
 941 if self . options . get ( 'as_recarray' ) : 

 /databricks/python/lib/python3.5/site-packages/pandas/io/parsers.py in read (self, nrows) 
 1506 def read ( self , nrows = None ) : 
 1507 try : 
 -> 1508 data = self . _reader . read ( nrows ) 
 1509 except StopIteration : 
 1510 if self . _first_chunk : 

 /databricks/python/lib/python3.5/site-packages/pandas/parser.cpython-35m-x86_64-linux-gnu.so in pandas.parser.TextReader.read (pandas/parser.c:9977) () 

 /databricks/python/lib/python3.5/site-packages/pandas/parser.cpython-35m-x86_64-linux-gnu.so in pandas.parser.TextReader._read_low_memory (pandas/parser.c:10235) () 

 /databricks/python/lib/python3.5/site-packages/pandas/parser.cpython-35m-x86_64-linux-gnu.so in pandas.parser.TextReader._read_rows (pandas/parser.c:11254) () 

 /databricks/python/lib/python3.5/site-packages/pandas/parser.cpython-35m-x86_64-linux-gnu.so in pandas.parser.TextReader._convert_column_data (pandas/parser.c:12629) () 

 /databricks/python/lib/python3.5/site-packages/pandas/parser.cpython-35m-x86_64-linux-gnu.so in pandas.parser.TextReader._convert_tokens (pandas/parser.c:13876) () 

 ValueError : could not convert string to float: '#DIV/0!'

The error indicates that `#DIV/0!` is a value in the `score_y` field. So this value is included in the `na_values` parameter (below).

In [33]:
%python
import numpy as np
mad_men_show_data_pdf = pd.read_csv('/dbfs/mnt/datalab-datasets/538_data/mad-men/show-data.csv', 
                                    names=['performer', 'show', 'show_start', 'show_end',
                                           'status', 'char_end', 'years_since', 'lead', 
                                           'support', 'shows', 'score', 'score_y',
                                           'lead_notes', 'support_notes', 'show_notes'],
                                    dtype={'show_start': np.float32, 
                                           'show_end': np.float32,
                                           'score_y': np.float32},
                                    header=0, 
                                    na_values=['PRESENT','#DIV/0!'],
                                    encoding = 'ISO-8859-1')
mad_men_show_data_pdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 248 entries, 0 to 247
Data columns (total 15 columns):
performer 248 non-null object
show 248 non-null object
show_start 248 non-null float32
show_end 219 non-null float32
status 248 non-null object
char_end 248 non-null int64
years_since 248 non-null int64
lead 248 non-null int64
support 248 non-null int64
shows 248 non-null int64
score 248 non-null float64
score_y 245 non-null float32
lead_notes 89 non-null object
support_notes 135 non-null object
show_notes 138 non-null object
dtypes: float32(3), float64(1), int64(5), object(6)
memory usage: 26.2+ KB

Great. All columns are read with acceptable datatypes. The following command displays the first few records for `object` columns.

In [35]:
%python
import numpy as np
mad_men_show_data_pdf = pd.read_csv('/dbfs/mnt/datalab-datasets/538_data/mad-men/show-data.csv', 
                                    names=['performer', 'show', 'show_start', 'show_end',
                                           'status', 'char_end', 'years_since', 'lead', 
                                           'support', 'shows', 'score', 'score_y',
                                           'lead_notes', 'support_notes', 'show_notes'],
                                    dtype={'show_start': np.float32, 
                                           'show_end': np.float32,
                                           'score_y': np.float32},
                                    header=0, 
                                    na_values=['PRESENT','#DIV/0!'],
                                    encoding = 'ISO-8859-1')
mad_men_show_data_pdf \
  .select_dtypes(include=['object']) \
  .head(5)

Out[ 49 ]: 
 performer show status \
0 Steven Hill Law & Order END 
1 Kelli Williams The Practice END 
2 LisaGay Hamilton The Practice END 
3 Lara Flynn Boyle The Practice END 
4 Dylan McDermott The Practice END 

 lead_notes \
0 NaN 
1 NaN 
2 Life of a King, 2014; Go For Sisters, 2013 
3 NaN 
4 Olympus Has Fallen, 2013; Freezer, 2014 

 support_notes \
0 NaN 
1 Any Day Now (2012) 
2 NaN 
3 NaN 
4 The Messengers, 2007; Unbeatable Harold, 2009;... 

 show_notes 
0 NaN 
1 Medical Investigation, Season 1; Lie To Me, Se... 
2 Men of a Certain Age, Season 1-2 
3 NaN 
4 Big Shots, Season 1; Dark Blue, Season 1-2; Am...

One final modification is to check on the `status` column. It is likely that it should have datatype of `category`.

First, check the unique values of this column.

In [37]:
mad_men_show_data_pdf.status.value_counts()

Out[ 50 ]: 
END 202
LEFT 29
End 17
Name: status, dtype: int64

Now change the datatype for the `status` column with the `dtype` parameter.

In [39]:
%python
import numpy as np
mad_men_show_data_pdf = pd.read_csv('/dbfs/mnt/datalab-datasets/538_data/mad-men/show-data.csv', 
                                    names=['performer', 'show', 'show_start', 'show_end',
                                           'status', 'char_end', 'years_since', 'lead', 
                                           'support', 'shows', 'score', 'score_y',
                                           'lead_notes', 'support_notes', 'show_notes'],
                                    dtype={'show_start': np.float32, 
                                           'show_end': np.float32,
                                           'score_y': np.float32,
                                           'status': 'category'},
                                    header=0, 
                                    na_values=['PRESENT','#DIV/0!'],
                                    encoding = 'ISO-8859-1')
mad_men_show_data_pdf \
  .info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 248 entries, 0 to 247
Data columns (total 15 columns):
performer 248 non-null object
show 248 non-null object
show_start 248 non-null float32
show_end 219 non-null float32
status 248 non-null category
char_end 248 non-null int64
years_since 248 non-null int64
lead 248 non-null int64
support 248 non-null int64
shows 248 non-null int64
score 248 non-null float64
score_y 245 non-null float32
lead_notes 89 non-null object
support_notes 135 non-null object
show_notes 138 non-null object
dtypes: category(1), float32(3), float64(1), int64(5), object(5)
memory usage: 24.6+ KB

Great. Now, all columns really are read with acceptable datatypes.

## 3. `read_excel`

List the spreadsheets in the `per_diem` directory.

In [43]:
%sh ls -hot /dbfs/mnt/datalab-datasets/per_diem/*.xls

-rw-r--r-- 1 root 257K Sep 18 12:15 /dbfs/mnt/datalab-datasets/per_diem/April2017PD.xls
-rw-r--r-- 1 root 242K Sep 18 12:15 /dbfs/mnt/datalab-datasets/per_diem/April2018PD.xls
-rw-r--r-- 1 root 246K Sep 18 12:15 /dbfs/mnt/datalab-datasets/per_diem/August2017PD.xls
-rw-r--r-- 1 root 238K Sep 18 12:15 /dbfs/mnt/datalab-datasets/per_diem/December2017PD.xls
-rw-r--r-- 1 root 263K Sep 18 12:15 /dbfs/mnt/datalab-datasets/per_diem/February2017PD.xls
-rw-r--r-- 1 root 240K Sep 18 12:15 /dbfs/mnt/datalab-datasets/per_diem/February2018PD.xls
-rw-r--r-- 1 root 251K Sep 18 12:15 /dbfs/mnt/datalab-datasets/per_diem/January2017PD.xls
-rw-r--r-- 1 root 233K Sep 18 12:15 /dbfs/mnt/datalab-datasets/per_diem/January2018PD.xls
-rw-r--r-- 1 root 241K Sep 18 12:15 /dbfs/mnt/datalab-datasets/per_diem/July2017PD.xls
-rw-r--r-- 1 root 259K Sep 18 12:15 /dbfs/mnt/datalab-datasets/per_diem/July2018PD.xls
-rw-r--r-- 1 root 250K Sep 18 12:15 /dbfs/mnt/datalab-datasets/per_diem/June2017PD.xls
-rw-r--r-- 1 root 253K Sep 18 12:15 /dbfs/mnt/datalab-datasets/per_diem/June2018PD.xls
-rw-r--r-- 1 root 249K Sep 18 12:15 /dbfs/mnt/datalab-datasets/per_diem/March2017PD.xls
-rw-r--r-- 1 root 242K Sep 18 12:15 /dbfs/mnt/datalab-datasets/per_diem/March2018PD.xls
-rw-r--r-- 1 root 257K Sep 18 12:15 /dbfs/mnt/datalab-datasets/per_diem/May2017PD.xls
-rw-r--r-- 1 root 256K Sep 18 12:15 /dbfs/mnt/datalab-datasets/per_diem/May2018PD.xls
-rw-r--r-- 1 root 237K Sep 18 12:15 /dbfs/mnt/datalab-datasets/per_diem/November2017PD.xls
-rw-r--r-- 1 root 237K Sep 18 12:15 /dbfs/mnt/datalab-datasets/per_diem/October2017PD.xls
-rw-r--r-- 1 root 242K Sep 18 12:15 /dbfs/mnt/datalab-datasets/per_diem/September2017PD.xls
-rw-r--r-- 1 root 245K Sep 18 12:15 /dbfs/mnt/datalab-datasets/per_diem/September2018PD.xls
-rw-r--r-- 1 root 253K Sep 18 12:15 /dbfs/mnt/datalab-datasets/per_diem/August2018PD.xls

The following command uses the pandas function `read_excel` to read in the spreadsheet.

In [45]:
%python
import pandas as pd
may_2017_pdf = pd.read_excel('/dbfs/mnt/datalab-datasets/per_diem/May2017PD.xls')
may_2017_pdf.info()

WARNING *** file size (262658) not 512 + multiple of sector size (512)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1105 entries, 0 to 1104
Data columns (total 11 columns):
Country Name 1105 non-null object
Location 1105 non-null object
Season Code 1105 non-null object
Season Start Date 1105 non-null datetime64[ns]
Season End Date 1105 non-null datetime64[ns]
Lodging 1105 non-null int64
Meals & Incidentals 1105 non-null int64
Per Diem 1105 non-null int64
Effective Date 1105 non-null datetime64[ns]
Footnote Reference 340 non-null object
Location Code 1105 non-null int64
dtypes: datetime64[ns](3), int64(4), object(4)
memory usage: 95.0+ KB

Notice the column names (they are in mixed case and contain spaces) and notice the columns with type `object`. 

We fix the column names with the `names` parameter.

In [47]:
%python
import pandas as pd
may_2017_pdf = pd.read_excel('/dbfs/mnt/datalab-datasets/per_diem/May2017PD.xls',
                             names=['country_name', 'location', 'season_code', 'season_start_date', 
                                    'season_end_date', 'lodging', 'meals_incidentals', 'per_diem', 
                                    'effective_date', 'footnote_reference', 'location_code']
                            )
may_2017_pdf.info()

WARNING *** file size (262658) not 512 + multiple of sector size (512)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1105 entries, 0 to 1104
Data columns (total 11 columns):
country_name 1105 non-null object
location 1105 non-null object
season_code 1105 non-null object
season_start_date 1105 non-null datetime64[ns]
season_end_date 1105 non-null datetime64[ns]
lodging 1105 non-null int64
meals_incidentals 1105 non-null int64
per_diem 1105 non-null int64
effective_date 1105 non-null datetime64[ns]
footnote_reference 340 non-null object
location_code 1105 non-null int64
dtypes: datetime64[ns](3), int64(4), object(4)
memory usage: 95.0+ KB

In [48]:
may_2017_pdf \
  .select_dtypes(include=['object']) \
  .head()

Out[ 7 ]: 
 country_name location season_code footnote_reference
0 AFGHANISTAN [Other] S1 19,2
1 AFGHANISTAN Kabul S1 19,2
2 ALBANIA [Other] S1 NaN
3 ALBANIA Tirana S1 NaN
4 ALGERIA [Other] S1 2

The first three columns should have datatype `category`. Make is so number 1.

In [50]:
%python
import pandas as pd
may_2017_pdf = pd.read_excel('/dbfs/mnt/datalab-datasets/per_diem/May2017PD.xls',
                             names=['country_name', 'location', 'season_code', 'season_start_date', 
                                    'season_end_date', 'lodging', 'meals_incidentals', 'per_diem', 
                                    'effective_date', 'footnote_reference', 'location_code'],
                             dtypes={'country_name': 'category',
                                     'location': 'category',
                                     'season_code': 'category'}
                            )
may_2017_pdf.info()

WARNING *** file size (262658) not 512 + multiple of sector size (512)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1105 entries, 0 to 1104
Data columns (total 11 columns):
country_name 1105 non-null object
location 1105 non-null object
season_code 1105 non-null object
season_start_date 1105 non-null datetime64[ns]
season_end_date 1105 non-null datetime64[ns]
lodging 1105 non-null int64
meals_incidentals 1105 non-null int64
per_diem 1105 non-null int64
effective_date 1105 non-null datetime64[ns]
footnote_reference 340 non-null object
location_code 1105 non-null int64
dtypes: datetime64[ns](3), int64(4), object(4)
memory usage: 95.0+ KB

Unfortunately, the `dtypes` parameter didn't work.

In [52]:
%python
import pandas as pd
may_2017_pdf = pd.read_excel('/dbfs/mnt/datalab-datasets/per_diem/May2017PD.xls',
                             names=['country_name', 'location', 'season_code', 'season_start_date', 
                                    'season_end_date', 'lodging', 'meals_incidentals', 'per_diem', 
                                    'effective_date', 'footnote_reference', 'location_code'],
                             dtype={'country_name': 'category',
                                     'location': 'category',
                                     'season_code': 'category'}
                            )
may_2017_pdf.info()

--------------------------------------------------------------------------- 
 ValueError Traceback (most recent call last)
 <command-858169> in <module> () 
 6 dtype={'country_name': 'category',
 7 'location' : 'category' , 
 ----> 8 'season_code': 'category'}
 9 )
 10 may_2017_pdf . info ( ) 

 /databricks/python/lib/python3.5/site-packages/pandas/io/excel.py in read_excel (io, sheetname, header, skiprows, skip_footer, index_col, names, parse_cols, parse_dates, date_parser, na_values, thousands, convert_float, has_index_names, converters, true_values, false_values, engine, squeeze, **kwds) 
 198 skip_footer = skip_footer , converters = converters , 
 199 true_values = true_values , false_values = false_values , squeeze = squeeze , 
 --> 200 **kwds)
 201 
 202 

 /databricks/python/lib/python3.5/site-packages/pandas/io/excel.py in _parse_excel (self, sheetname, header, skiprows, names, skip_footer, index_col, has_index_names, parse_cols, parse_dates, date_parser, na_values, thousands, convert_float, true_values, false_values, verbose, squeeze, **kwds) 
 502 skipfooter = skip_footer , 
 503 squeeze = squeeze , 
 --> 504 **kwds)
 505 
 506 output [ asheetname ] = parser . read ( ) 

 /databricks/python/lib/python3.5/site-packages/pandas/io/parsers.py in TextParser (*args, **kwds) 
 1669 """
 1670 kwds [ 'engine' ] = 'python' 
 -> 1671 return TextFileReader ( * args , ** kwds ) 
 1672 
 1673 

 /databricks/python/lib/python3.5/site-packages/pandas/io/parsers.py in __init__ (self, f, engine, **kwds) 
 718 self . _currow = 0 
 719 
 --> 720 options = self . _get_options_with_defaults ( engine ) 
 721 
 722 self . chunksize = options . pop ( 'chunksize' , None ) 

 /databricks/python/lib/python3.5/site-packages/pandas/io/parsers.py in _get_options_with_defaults (self, engine) 
 759 raise ValueError(
 760 'The %r option is not supported with the' 
 --> 761 ' %r engine' % (argname, engine))
 762 else : 
 763 value = default 

 ValueError : The 'dtype' option is not supported with the 'python' engine

Checking this message on Google reveals that we should use the `converters` parameter. See [stackoverflow](https://stackoverflow.com/questions/32591466/python-pandas-how-to-specify-data-types-when-reading-an-excel-file).

Further checking [the documentation](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_excel.html) reveals that the `dtype` parameter was added in version `0.20.0` and we are running `0.19.2`. (See below.)

In [54]:
pd.__version__

Out[ 17 ]: '0.19.2'

There seems to be little documentation on the `converters` parameter.  Because of this I use the `assign` method to change the type of the column after reading the spreadsheet with `read_excel` as follows.

In [56]:
%python
import pandas as pd
may_2017_pdf = \
pd.read_excel('/dbfs/mnt/datalab-datasets/per_diem/May2017PD.xls',
              names=['country_name', 'location', 'season_code', 'season_start_date', 
                     'season_end_date', 'lodging', 'meals_incidentals', 'per_diem', 
                     'effective_date', 'footnote_reference', 'location_code']
             ) \
  .assign(country_name=lambda df: df.country_name.astype('category'),
          location    =lambda df: df.location.astype('category'),
          season_code =lambda df: df.season_code.astype('category')
         )
        
may_2017_pdf.info()

WARNING *** file size (262658) not 512 + multiple of sector size (512)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1105 entries, 0 to 1104
Data columns (total 11 columns):
country_name 1105 non-null category
location 1105 non-null category
season_code 1105 non-null category
season_start_date 1105 non-null datetime64[ns]
season_end_date 1105 non-null datetime64[ns]
lodging 1105 non-null int64
meals_incidentals 1105 non-null int64
per_diem 1105 non-null int64
effective_date 1105 non-null datetime64[ns]
footnote_reference 340 non-null object
location_code 1105 non-null int64
dtypes: category(3), datetime64[ns](3), int64(4), object(1)
memory usage: 83.2+ KB

See the documentation on the `assign` method:
- https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.assign.html

Make one last check on the columns of type `object`.

In [59]:
%python
may_2017_pdf \
  .select_dtypes(include=['object']) \
  .head()

Out[ 28 ]: 
 footnote_reference
0 19,2
1 19,2
2 NaN
3 NaN
4 2

That looks like a reasonable `object` column. The `may_2017_pdf` dataframe is ready to be analyzed.

## 4. `read_json`

List the JSON files in the `/dbfs/mnt/datalab-datasets/JSON` directory.

In [63]:
%sh ls -hot /dbfs/mnt/datalab-datasets/JSON

total 0
-rw-r--r-- 1 root 2.9M Aug 7 2017 world_bank.json
-rw-r--r-- 1 root 3.1M Aug 7 2017 zips.json
-rw-r--r-- 1 root 11M Aug 7 2017 stocks.json
-rw-r--r-- 1 root 16M Aug 7 2017 enron.json
-rw-r--r-- 1 root 75M Aug 7 2017 companies.json

The first example uses the `zips.json` file. It only contains three records, which are displayed below.

In [65]:
%sh head -n 3 /dbfs/mnt/datalab-datasets/JSON/zips.json

{ "city" : "AGAWAM", "loc" : [ -72.622739, 42.070206 ], "pop" : 15338, "state" : "MA", "_id" : "01001" }
{ "city" : "CUSHMAN", "loc" : [ -72.51564999999999, 42.377017 ], "pop" : 36963, "state" : "MA", "_id" : "01002" }
{ "city" : "BARRE", "loc" : [ -72.10835400000001, 42.409698 ], "pop" : 4546, "state" : "MA", "_id" : "01005" }

Read the file using `read_json`. The `lines=True` parameter indicates that there should be one JSON object per line. These JSON objects correspond to rows/records. The key names correspond to column names.

In [67]:
%python
zips_pdf = pd.read_json('/dbfs/mnt/datalab-datasets/JSON/zips.json',
                        lines=True)
zips_pdf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 29467 entries, 0 to 29466
Data columns (total 5 columns):
_id 29467 non-null int64
city 29467 non-null object
loc 29467 non-null object
pop 29467 non-null int64
state 29467 non-null object
dtypes: int64(2), object(3)
memory usage: 1.3+ MB

List the first five records of the dataframe. Notice the `loc` column. It seems to contain a list, but has the `object` datatype.

In [69]:
zips_pdf.head()

Out[ 41 ]: 
 _id city loc pop state
0 1001 AGAWAM [-72.622739, 42.070206] 15338 MA
1 1002 CUSHMAN [-72.51565, 42.377017] 36963 MA
2 1005 BARRE [-72.108354, 42.409698] 4546 MA
3 1007 BELCHERTOWN [-72.410953, 42.275103] 10579 MA
4 1008 BLANDFORD [-72.936114, 42.182949] 1240 MA

In [70]:
%python 
zips_pdf.loc

Out[ 42 ]: <pandas.core.indexing._LocIndexer at 0x7fc17ad69c18>

The `loc` indexer masks the `loc column` so it should be renamed. The `read_json` function has no `names` parameter so we use the `rename` method. See the documentation for details on the `rename` method: 
- https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.rename.html

In [72]:
%python
zips_pdf = \
pd.read_json('/dbfs/mnt/datalab-datasets/JSON/zips.json',
             lines=True) \
  .rename(columns={'loc':'lon_lat'})
zips_pdf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 29467 entries, 0 to 29466
Data columns (total 5 columns):
_id 29467 non-null int64
city 29467 non-null object
lon_lat 29467 non-null object
pop 29467 non-null int64
state 29467 non-null object
dtypes: int64(2), object(3)
memory usage: 1.3+ MB

The `city` and `state` columns should have datatype `category`. The `dtype` parameter didn't work so the `assign` method is used.

In [74]:
%python
zips_pdf = \
pd.read_json('/dbfs/mnt/datalab-datasets/JSON/zips.json',
             lines=True) \
  .rename(columns={'loc':'lon_lat'}) \
  .assign(city =lambda df: df.city .astype('category'),
          state=lambda df: df.state.astype('category')
         )

zips_pdf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 29467 entries, 0 to 29466
Data columns (total 5 columns):
_id 29467 non-null int64
city 29467 non-null category
lon_lat 29467 non-null object
pop 29467 non-null int64
state 29467 non-null category
dtypes: category(2), int64(2), object(1)
memory usage: 1.1+ MB

The `lon_lat` column should be split into two columns, but that will have to wait until later.

__The End__